src/client/cli/cmd/shell.cpp

```diff
@@ -37,11 +42,24 @@ mp::ReturnCode cmd::Shell::run(mp::ArgParser* parser)
         return parser->returnCodeFrom(ret);
     }
 
+    std::unique_ptr<mp::utils::Timer> timer;
+
+    if (parser->isSet("timeout"))
+    {
+        timer = std::make_unique<mp::utils::Timer>(std::chrono::seconds(parser->value("timeout").toInt()), []() {
+            std::cerr << "Timed out waiting for instance to start." << std::endl;
+            std::raise(SIGINT);
+        });
+    }
+
     // We can assume the first array entry since `shell` only uses one instance
     // at a time
     auto instance_name = request.instance_name()[0];
 
-    auto on_success = [this](mp::SSHInfoReply& reply) {
+    auto on_success = [this, &timer](mp::SSHInfoReply& reply) {
+        if (timer)
+            timer->stop();
```

Comment:
And this is uncovered because we've currently no way to wait for the timeout to pass below - for testing it just returns `Ok` straight up.